src/Mono.Android/Android.Runtime/JNIEnv.cs

```diff
@@ -1590,13 +1590,40 @@ public static IntPtr NewArray (Array value, Type? elementType = null)
 			return creator (value);
 		}
 
+		static IntPtr NewArray (Array value, Type elementType, IntPtr elementClass)
+		{
+			if (value == null)
+				throw new ArgumentNullException ("value");
+
+			IntPtr array = IntPtr.Zero;
+			try {
+				array = NewObjectArray (value.Length, elementClass, IntPtr.Zero);
+
+				for (int i = 0; i < value.Length; ++i) {
+					IntPtr subarray = NewArray ((Array) value.GetValue (i)!, elementType.GetElementType ());
+					SetObjectArrayElement (array, i, subarray);
+					DeleteLocalRef (subarray);
+				}
+
+				return array;
+			} catch {
+				DeleteLocalRef (array);
+				throw;
+			}
+		}
+
 		public static IntPtr NewArray<T> (T[]? array)
 		{
 			if (array == null)
 				return IntPtr.Zero;
 
 			if (typeof (T).IsArray) {
-				return NewArray (array, typeof (T));
+				IntPtr elementClass = FindClass (typeof (T));
+				try {
+					return NewArray (array, typeof (T), elementClass);
```

Comment:
@jonpryor can we really do this for any type `T`? This seems safe for `int[][]`, for example, but would `Java.Lang.Object[][]` have an issue if individual array items are different types?

Reply:
@jonathanpeppers: generally, yes, we can do this for any type `T`.  `JNIEnv.NewArray<T>(T[])` is for "compile-time" knowledge, not runtime knowledge

There is an "implicit" dependency that the type created by `JNIEnv.NewArray()` be compatible with the compile-time type of the declared method: if the Java method wants an `int[][]`, it better *get* an `int[][]`, or you'll have a runtime exception.

A Java method could certainly require additional runtime types, e.g. it accepts a `Object[]`, and it treats each element differently based on the runtime type, but this is a scenario that hasn't bitten us yet. ;-)

*Most*  multidimensional array usage in `Mono.Android.dll` is for *non*-`Object` types:

```
% grep -ri '\(public\|protected\).*\[\]\[\]' src/Mono.Android/obj/Debug/monoandroid10
src/Mono.Android/obj/Debug/monoandroid10/android-32/mcw/Android.Icu.Util.Calendar.cs:		protected virtual unsafe int[][][]? GetFieldResolutionTable ()
src/Mono.Android/obj/Debug/monoandroid10/android-32/mcw/Android.Icu.Util.Calendar.cs:		protected virtual unsafe int ResolveFields (int[][][]? precedenceTable)
src/Mono.Android/obj/Debug/monoandroid10/android-32/mcw/Java.Util.ListResourceBundle.cs:		protected abstract Java.Lang.Object[][]? GetContents ();
src/Mono.Android/obj/Debug/monoandroid10/android-32/mcw/Java.Util.ListResourceBundle.cs:		protected override unsafe Java.Lang.Object[][]? GetContents ()
src/Mono.Android/obj/Debug/monoandroid10/android-32/mcw/Android.Animation.PropertyValuesHolder.cs:		public static unsafe Android.Animation.PropertyValuesHolder? OfMultiFloat (string? propertyName, float[][]? values)
src/Mono.Android/obj/Debug/monoandroid10/android-32/mcw/Android.Animation.PropertyValuesHolder.cs:		public static unsafe Android.Animation.PropertyValuesHolder? OfMultiInt (string? propertyName, int[][]? values)
src/Mono.Android/obj/Debug/monoandroid10/android-32/mcw/Android.Nfc.NfcAdapter.cs:		public unsafe void EnableForegroundDispatch (Android.App.Activity? activity, Android.App.PendingIntent? intent, Android.Content.IntentFilter[]? filters, string[][]? techLists)
src/Mono.Android/obj/Debug/monoandroid10/android-32/mcw/Javax.Security.Auth.PrivateCredentialPermission.cs:		public unsafe string[][]? GetPrincipals ()
src/Mono.Android/obj/Debug/monoandroid10/android-32/mcw/Android.Net.SSLCertificateSocketFactory.cs:		public virtual unsafe void SetNpnProtocols (byte[][]? npnProtocols)
src/Mono.Android/obj/Debug/monoandroid10/android-32/mcw/Android.Locations.GnssAntennaInfo.cs:			public unsafe SphericalCorrections (double[][] corrections, double[][] correctionUncertainties) : base (IntPtr.Zero, JniHandleOwnership.DoNotTransfer)
src/Mono.Android/obj/Debug/monoandroid10/android-32/mcw/Android.Locations.GnssAntennaInfo.cs:			public unsafe double[][] GetCorrectionsArray ()
src/Mono.Android/obj/Debug/monoandroid10/android-32/mcw/Android.Locations.GnssAntennaInfo.cs:			public unsafe double[][] GetCorrectionUncertaintiesArray ()
src/Mono.Android/obj/Debug/monoandroid10/android-32/mcw/Java.Text.DateFormatSymbols.cs:		public virtual unsafe string[][]? GetZoneStrings ()
src/Mono.Android/obj/Debug/monoandroid10/android-32/mcw/Java.Text.DateFormatSymbols.cs:		public virtual unsafe void SetZoneStrings (string[][]? newZoneStrings)
src/Mono.Android/obj/Debug/monoandroid10/android-32/mcw/Java.Lang.Reflect.Method.cs:		public override unsafe Java.Lang.Annotation.IAnnotation[][] GetParameterAnnotations ()
src/Mono.Android/obj/Debug/monoandroid10/android-32/mcw/Android.Content.Res.ColorStateList.cs:		public unsafe ColorStateList (int[][]? states, int[]? colors) : base (IntPtr.Zero, JniHandleOwnership.DoNotTransfer)
src/Mono.Android/obj/Debug/monoandroid10/android-32/mcw/Java.Lang.Reflect.Executable.cs:		public abstract Java.Lang.Annotation.IAnnotation[][] GetParameterAnnotations ();
src/Mono.Android/obj/Debug/monoandroid10/android-32/mcw/Java.Lang.Reflect.Executable.cs:		public override unsafe Java.Lang.Annotation.IAnnotation[][] GetParameterAnnotations ()
src/Mono.Android/obj/Debug/monoandroid10/android-32/mcw/Android.Icu.Text.DateFormatSymbols.cs:		public virtual unsafe string[][]? GetZoneStrings ()
src/Mono.Android/obj/Debug/monoandroid10/android-32/mcw/Android.Icu.Text.DateFormatSymbols.cs:		public virtual unsafe void SetZoneStrings (string[][]? newZoneStrings)
src/Mono.Android/obj/Debug/monoandroid10/android-32/mcw/Android.Animation.ObjectAnimator.cs:		public static unsafe Android.Animation.ObjectAnimator? OfMultiFloat (Java.Lang.Object? target, string? propertyName, float[][]? values)
src/Mono.Android/obj/Debug/monoandroid10/android-32/mcw/Android.Animation.ObjectAnimator.cs:		public static unsafe Android.Animation.ObjectAnimator? OfMultiInt (Java.Lang.Object? target, string? propertyName, int[][]? values)
src/Mono.Android/obj/Debug/monoandroid10/android-32/mcw/Android.App.AppSearch.GenericDocument.cs:			public virtual unsafe Java.Lang.Object SetPropertyBytes (string name, params byte[][] values)
src/Mono.Android/obj/Debug/monoandroid10/android-32/mcw/Android.App.AppSearch.GenericDocument.cs:		public virtual unsafe byte[][]? GetPropertyBytesArray (string path)
src/Mono.Android/obj/Debug/monoandroid10/android-32/mcw/Java.Lang.Reflect.Constructor.cs:		public override unsafe Java.Lang.Annotation.IAnnotation[][]? GetParameterAnnotations ()
```

Only [`ListResourceBundle.GetContents()`](https://developer.android.com/reference/java/util/ListResourceBundle) deals with `Object[][]`, and it doesn't have any further runtime requirements.